SparkOperator: Prometheus Metrics Integration

akhurana001 opened the pull request:
https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/67

liyinan926 commented on docs/quick-start-guide.md:
```diff
@@ -50,6 +50,41 @@ $ kubectl describe deployment sparkoperator -n sparkoperator
 
+### Metrics
+
+The operator exposes a set of metrics via the metric endpoint to be scraped by `Prometheus`. This is enabled by default and can be disabled by setting  the flag `-enable-metrics=false` and running 
```
I'm thinking maybe the this should default to `false` since not everyone is running Prometheus.

akhurana001 replied:
done

liyinan926 commented on main.go:
```diff
@@ -20,29 +20,28 @@ package main
 	sacrd "k8s.io/spark-on-k8s-operator/pkg/crd/sparkapplication"
+	"k8s.io/spark-on-k8s-operator/pkg/util"
 	"k8s.io/spark-on-k8s-operator/pkg/webhook"
+	"os"
```
Built-ion packages should come at the top.

akhurana001 replied:
ran goimports on this and other files. I was using gofmt earlier which ended up reworking the imports in alphabetical order.

liyinan926 commented on main.go:
```diff
@@ -73,6 +80,12 @@ func main() {
 
+	var metricsBundle *util.PrometheusMetrics
+	if *enableMetrics {
+		glog.Info("Enabling metrics")
```
`enabling metrics collecting and exporting to Prometheus`.

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -523,17 +567,149 @@ func (c *Controller) updateSparkApplicationStatusWithRetries(
+
+// isValidExecutorStatusTransition determines if the executor state transition is valid
+func isValidExecutorStatusTransition(oldStatus, newStatus v1alpha1.ExecutorState) bool {
+
```
Please remove the extra empty line here and in several places below.

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -55,6 +53,38 @@ const (
 var (
 	keyFunc = cache.DeletionHandlingMetaNamespaceKeyFunc
+
+	// Valid Driver State Transitions: map with valid pairs of oldState-> {newStates}
```
I think it's worth putting code related to state transition validation into a separate go file, e.g., state_validation.go?

akhurana001 replied:
sure, makes sense.

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -500,21 +532,33 @@ func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdat
-	updateFunc func(*v1alpha1.SparkApplicationStatus)) *v1alpha1.SparkApplication {
-	toUpdate := original.DeepCopy()
+	updateFunc func(*v1alpha1.SparkApplicationStatus),
+) *v1alpha1.SparkApplication {
```
You can revert this line. It doesn't need to be in a new line.

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -500,21 +532,33 @@ func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdat
 			return updated
 		}
 		lastUpdateErr = err
+		glog.Errorf("[Attempt: %v] failed to update SparkApplication [%v] %v", i, toUpdate, err)
```
The first `%v` can be `%d`. Also do you really want to log the entire `SparkApplication` object instead of just the app name?

akhurana001 replied:
I wanted to log more details since this is an error state. For now, I changed this to just log the name. Will probably add more details once we have the state management change out and this should be pretty rare

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -500,21 +532,33 @@ func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdat
 	var lastUpdateErr error
 	for i := 0; i < maximumUpdateRetries; i++ {
-		updated, err := c.tryUpdateStatus(original, toUpdate, updateFunc)
+		toUpdate := original.DeepCopy()
```
You only need to make a deep copy once outside the loop. You can assign `toUpdate` to is fetched from the API server below as the one returned is a copy anyway.

akhurana001 replied:
done

akhurana001 replied:
We actually need this and I have reverted it back to make a deep copy inside the loop (https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/pull/227/commits/f600336b353287b0cd983148071c2bf9640b1135)

This is because for metrics,  we need the oldState of the CRD which we actually updated and since we run `updateFunc(&toUpdate.Status)` , `toUpdate` is no longer the oldState which was read from the API server nor is `original`. Let me know if you have any questions.

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -523,17 +567,149 @@ func (c *Controller) updateSparkApplicationStatusWithRetries(
+	return false
+}
+
+func fetchMetricLabels(specLabels map[string]string, labels []string) map[string]string {
```
Let's put metrics related code into its own file, e.g., metrics.go.

akhurana001 replied:
sure, refactored as below and moved this out

liyinan926 commented on pkg/controller/sparkapplication/controller_test.go:
```diff
@@ -32,9 +31,13 @@ import (
 	crdclientfake "k8s.io/spark-on-k8s-operator/pkg/client/clientset/versioned/fake"
 	crdinformers "k8s.io/spark-on-k8s-operator/pkg/client/informers/externalversions"
+	"k8s.io/spark-on-k8s-operator/pkg/util"
+	"net/http"
```
Built-in package should be at the top.

liyinan926 commented on pkg/controller/sparkapplication/controller_test.go:
```diff
@@ -17,13 +17,12 @@ limitations under the License.
 package sparkapplication
 
 import (
+	"github.com/stretchr/testify/assert"
```
Please revert this position change.

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -523,17 +567,149 @@ func (c *Controller) updateSparkApplicationStatusWithRetries(
+
+func (c *Controller) postMetrics(oldState, newState *v1alpha1.SparkApplication) {
+	if c.metrics == nil {
+		glog.V(2).Infof("Metrics not enabled. Skipping posting metrics")
```
`Metrics not enabled so skipping exporting metrics`. The same below.

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -523,17 +567,149 @@ func (c *Controller) updateSparkApplicationStatusWithRetries(
+
+func fetchMetricLabels(specLabels map[string]string, labels []string) map[string]string {
+
+	metricLabels := map[string]string{}
```
Please use `make(map[string]string)` to be consistent with the rest of the code base.

akhurana001 replied:
done

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,250 @@
+package util
```
Missing license header.

akhurana001 replied:
done, here and other files as well

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -523,17 +567,149 @@ func (c *Controller) updateSparkApplicationStatusWithRetries(
+	return metricLabels
+}
+
+func (c *Controller) postMetrics(oldState, newState *v1alpha1.SparkApplication) {
```
`exportMetrics`, and let's rename `oldState` and `newState` to `oldApp` and `newApp`.

akhurana001 replied:
done

akhurana001 commented:
Hi, I have updated based on the PR comments.  I have refactored metrics implementation by moving the SparkApplication specific logic under  pkg/sparkapplication/sparkapp_metrics.go and only keeping the common logic in util/metrics.go.  Also updated the state-transition map to let terminal states move to Submitted as we support RetryPolicy of Always.

A couple of other pointers:
Based on my testing, the executor running count is off because the operator doesn't seem to be getting all executor events. I am guessing this is due to the driver pod/spark-distribution itself killing the executors on completion ? We can potentially mark executors as done if we get a terminal state for a driver. Thoughts ? 

Also I was trying to run a FailedSubmission retry example and looks like we might not be enforcing retryAttempts count on submission failure potentially due to race-conditions in state-management . I don't think that is related to this PR and I will work on a fix for that in https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/issues/199 . Let me know if you feel this PR can cause this behavior and I can take a deeper look.

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,233 @@
+}
+
+func (sm *SparkAppMetrics) exportMetrics(oldApp, newApp *v1alpha1.SparkApplication) {
+
```
This line can be removed.

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,233 @@
+	metricLabels := fetchMetricLabels(newApp.Labels, sm.Labels)
+	glog.V(2).Infof("Posting Metrics for %s. OldStatus: %v NewStatus: %v", newApp.GetName(), oldApp.Status, newApp.Status)
+
+	oldStatus := oldApp.Status.AppState.State
```
Can you rename them to `oldState` and `newState`, respectively?

akhurana001 replied:
done

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,184 @@
+
+// Gauge with conditional decrement ensuring its value is never negative.
+func NewPositiveGauge(name string, description string, labels []string) *PositiveGauge {
+
```
This line can be removed.

akhurana001 replied:
done

liyinan926 commented on docs/quick-start-guide.md:
```diff
@@ -50,6 +50,41 @@ $ kubectl describe deployment sparkoperator -n sparkoperator
+ `kubectl apply -f manifest/spark-operator.yaml`. You will also need to add the prometheus annotations `prometheus.io/scrape`
+  and `prometheus.io/port` which are used by Prometheus to scrape the metric endpoint if you are enabling metrics.
+  
+If enabled, the operator generates the following metrics:
```
I suggest we use a table that explains what each metric is for.

akhurana001 replied:
makes sense, done

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,233 @@
+		},
+		labels,
+	)
+	sparkAppRunningCount := util.NewPositiveGauge(util.CreateValidMetricNameLabel(prefix, "spark_app_running_count"),
```
Why are these two gauges, whereas the other app-related metrics are `CounterVec`?

akhurana001 replied:
Counters can only be incremented while Gauges can be both incremented/decremented hence for running count, we need Gauges

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -501,15 +508,25 @@ func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdat
+		}
+		updated, err := c.tryUpdateStatus(original, toUpdate)
 		if err == nil {
+			c.postMetrics(original, updated)
```
We can add a condition around `postMetrics`, i.e., if `updated != nil` we will do `postMetrics`.

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -523,17 +540,30 @@ func (c *Controller) updateSparkApplicationStatusWithRetries(
 
+func (c *Controller) postMetrics(oldApp, newApp *v1alpha1.SparkApplication) {
+
+	if c.metrics == nil {
```
Like mentioned above, these two conditions can be moved out of this function.

akhurana001 replied:
removed this function

liyinan926 commented:
> Based on my testing, the executor running count is off because the operator doesn't seem to be getting all executor events. I am guessing this is due to the driver pod/spark-distribution itself killing the executors on completion ? We can potentially mark executors as done if we get a terminal state for a driver. Thoughts ?

What if you collect executor metrics in `SparkPodMonitor` instead of when updating application status?

liyinan926 commented:
> Also I was trying to run a FailedSubmission retry example and looks like we might not be enforcing retryAttempts count on submission failure potentially due to race-conditions in state-management . I don't think that is related to this PR and I will work on a fix for that in #199 . Let me know if you feel this PR can cause this behavior and I can take a deeper look.

Let's not worry about that in this PR. I will also take a look at that issue.

kumare3 commented:
@liyinan926 I agree that we should maintain readability and quality of code. Could we achieve this through something like a linter? https://github.com/alecthomas/gometalinter

We could set that up as the CI process

liyinan926 commented:
@kumare3 making style and correctness checks part of the CI process using tools like `go vet` or `go lint` sounds like a great idea.

kumare3 commented:
@liyinan926 since you are the owner of the repo, can you setup something, and we can all follow

liyinan926 commented:
@kumare3 Sure.

liyinan926 commented:
@kumare3 @akhurana001 I will try setting something up for future PRs. For this PR, thanks for fixing the styles and formatting manually. The code style should look good to me after the fixes. Thanks!

akhurana001 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,233 @@
+	util.RegisterMetric(sparkAppExecutorFailureCount)
+	util.RegisterMetric(sparkAppExecutorSuccessCount)
+
+	metricBundle := &SparkAppMetrics{
```
done

akhurana001 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,233 @@
+}
+
+func fetchMetricLabels(specLabels map[string]string, labels []string) map[string]string {
+
```
done

akhurana001 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,233 @@
+	"k8s.io/spark-on-k8s-operator/pkg/util"
+)
+
+type SparkAppMetrics struct {
```
makes sense, done

akhurana001 commented on pkg/util/metrics_test.go:
```diff
@@ -0,0 +1,70 @@
+)
+
+func TestPositiveGauge_EmptyLabels(t *testing.T) {
+
```
removed

akhurana001 commented on pkg/util/metrics_test.go:
```diff
@@ -0,0 +1,70 @@
+}
+
+func TestPositiveGauge_WithLabels(t *testing.T) {
+
```
removed

akhurana001 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,233 @@
+}
+
+func NewSparkAppMetrics(prefix string, labels []string) *SparkAppMetrics {
+
```
done

akhurana001 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -148,7 +157,6 @@ func (c *Controller) Start(workers int, stopCh <-chan struct{}) error {
 		// the worker after one second.
 		go wait.Until(c.runWorker, time.Second, stopCh)
 	}
-
```
added back

akhurana001 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -23,7 +23,6 @@ import (
 	"time"
 
 	"github.com/golang/glog"
-
```
done

akhurana001 commented on main.go:
```diff
@@ -20,28 +20,28 @@ package main
 	"time"
 
 	"github.com/golang/glog"
-
```
done

akhurana001 commented:
> Based on my testing, the executor running count is off because the operator doesn't seem to be getting all executor events. I am guessing this is due to the driver pod/spark-distribution itself killing the executors on completion ? We can potentially mark executors as done if we get a terminal state for a driver. Thoughts ?
> 
> What if you collect executor metrics in SparkPodMonitor instead of when updating application status?

Based on my debugging,  we are getting Executor Pod Deleted updates with podStatus still as Pending/Running which is why in  both SparkApplication Status and metrics, we see some of the executors stuck as pending/running. I am guessing this happens because on completion, the driver deletes the executors which might still be pending/running ? 
This can potentially be handled similar to the way we default to `PodFailed` state on Driver Deletion in https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/pkg/controller/sparkapplication/spark_pod_monitor.go#L164

If this sounds ok, I can test this a little more and create a PR with fix for both Executor Status in SparkApplication as well as metrics.

akhurana001 commented:
Also FYI, I created a separate `spark-operator-with-metrics.yaml` with all the additional configurations required to set-up the metrics. 

Since we don't want to turn these on by default, I reverted the old `spark-operator.yaml` to not have any metric related configs.

mrow4a commented:
Good idea, this will also play well with Helm

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,229 @@
+
+	metricLabels := make(map[string]string)
+	for _, label := range labels {
+		metricLabels[label] = validSpecLabels[label]
```
It's better to do the following instead:
```
if value, ok := validSpecLabels[label]; ok {
  metricLabels[label] = value
} else {
  metricLabels[label] = "Unknown"
}
```

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,229 @@
+
+func (sm *sparkAppMetrics) exportMetrics(oldApp, newApp *v1alpha1.SparkApplication) {
+	metricLabels := fetchMetricLabels(newApp.Labels, sm.Labels)
+	glog.V(2).Infof("Posting Metrics for %s. OldStatus: %v NewStatus: %v", newApp.GetName(), oldApp.Status, newApp.Status)
```
Can we change all occurrences of `posting` to `exporting`?

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,229 @@
+
+	// Potential Executor status updates
+	for executor, newExecState := range newApp.Status.ExecutorState {
+		if newExecState == v1alpha1.ExecutorRunningState && (isValidExecutorStateTransition(oldApp.Status.ExecutorState[executor], newExecState)) {
```
Can we use a `switch` block that checks `newExecState` instead of embedding the check into the if blocks?

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics_test.go:
```diff
@@ -0,0 +1,64 @@
+	assert.Equal(t, float64(10), fetchCounterValue(metrics.SparkAppExecutorSuccessCount, app1))
+	assert.Equal(t, float64(5), metrics.SparkAppExecutorRunningCount.Value(app1))
+	assert.Equal(t, float64(5), metrics.SparkAppRunningCount.Value(app1))
+
```
This empty line can be removed.

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/state_validation.go:
```diff
@@ -0,0 +1,78 @@
+		v1alpha1.CompletedState: {v1alpha1.SubmittedState},
+	}
+
+	// Valid Executor State Transitions: map with valid paris of oldState-> {newStates}
```
Typo: `paris` -> `pairs`.

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/state_validation.go:
```diff
@@ -0,0 +1,78 @@
+		v1alpha1.FailedSubmissionState: {v1alpha1.UnknownState, v1alpha1.SubmittedState,
+			v1alpha1.RunningState, v1alpha1.CompletedState, v1alpha1.FailedState},
+
+		v1alpha1.SubmittedState: {v1alpha1.RunningState, v1alpha1.CompletedState, v1alpha1.FailedState},
```
Can we also use empty lines to separate the following entries out for consistency?

akhurana001 replied:
done

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,182 @@
+// Best effort metric registration with Prometheus.
+func RegisterMetric(metric prometheus.Collector) {
+	if err := prometheus.Register(metric); err != nil {
+		glog.Warningf("Error while registering Prometheus metric: [%v]", err)
```
You might want to use `Errorf` instead.

akhurana001 replied:
done

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,182 @@
+}
+
+func InitializeMetrics(endpoint string, port string, prefix string) {
+
```
This empty line can be removed.

akhurana001 replied:
done

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,182 @@
+func (p *WorkQueueMetrics) NewDepthMetric(name string) workqueue.GaugeMetric {
+	depthMetric := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: CreateValidMetricNameLabel(p.prefix, name+"_depth"),
+		Help: "Current depth of workqueue: " + name,
```
Can we use `fmt.Sprintf` for formatting the `Help` strings?

akhurana001 replied:
done

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,182 @@
+
+// Decrement the metric only if its positive for the labels specified
+func (c *PositiveGauge) Dec(labelMap map[string]string) {
+
```
This empty line can be removed.

akhurana001 replied:
done

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,182 @@
+
+// Increment the Metric for the labels specified
+func (c *PositiveGauge) Inc(labelMap map[string]string) {
+
```
This empty line can be removed.

akhurana001 replied:
done

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,182 @@
+
+	if m, err := c.gaugeMetric.GetMetricWith(labelMap); err != nil {
+		glog.Errorf("Error while posting metrics: %v", err)
+
```
This empty line can be removed.

akhurana001 replied:
done

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,182 @@
+	// Start the metrics endpoint for Prometheus to scrape
+	http.Handle(endpoint, promhttp.Handler())
+	go http.ListenAndServe(port, nil)
+	glog.Infof("Started Metrics server at localhost%s%s", port, endpoint)
```
Should it be `"Started metrics server at %s:%s", endpoint, port)`?

akhurana001 replied:
So the endpoint is actually just the path. Right now this log line looks like `Started Metrics server at localhost:10254/metrics` which looks easy to understand  to me. 

Also got rid of `:` from user-passed port and explicitly appending it to the user-specified port.

liyinan926 commented:
> Based on my debugging, we are getting Executor Pod Deleted updates with podStatus still as Pending/Running which is why in both SparkApplication Status and metrics, we see some of the executors stuck as pending/running. I am guessing this happens because on completion, the driver deletes the executors which might still be pending/running? This can potentially be handled similar to the way we default to PodFailed state on Driver Deletion in https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/pkg/controller/sparkapplication/spark_pod_monitor.go#L164

Let me think about a bit more on this.

liyinan926 commented:
@akhurana001 Please rebase onto the latest master to resolve the conflict.

akhurana001 commented on main.go:
```diff
@@ -108,7 +122,7 @@ func main() {
 		factoryOpts...)
 	applicationController := sparkapplication.NewController(
-		crdClient, kubeClient, apiExtensionsClient, factory, *submissionRunnerThreads, *namespace)
+		crdClient, kubeClient, apiExtensionsClient, factory, *submissionRunnerThreads, *enableMetrics, *metricsPrefix, metricsLabels, *namespace)
```
makes sense, done

akhurana001 commented on pkg/controller/sparkapplication/controller_test.go:
```diff
@@ -287,10 +296,18 @@ func TestOnDelete(t *testing.T) {
 }
 
 func TestProcessSingleDriverStateUpdate(t *testing.T) {
+
```
done

akhurana001 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,229 @@
+	sparkAppSuccessExecutionTime := prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_success_execution_time_microseconds"),
+			Help: "Spark App Runtime via the Operator",
```
done

akhurana001 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,229 @@
+	sparkAppFailureExecutionTime := prometheus.NewSummaryVec(
+		prometheus.SummaryOpts{
+			Name: util.CreateValidMetricNameLabel(prefix, "spark_app_failure_execution_time_microseconds"),
+			Help: "Spark App Runtime via the Operator",
```
done

akhurana001 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,229 @@
+type sparkAppMetrics struct {
+	Labels               []string
+	prefix               string
+	SparkAppSubmitCount  *prometheus.CounterVec
```
done

akhurana001 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,229 @@
+)
+
+type sparkAppMetrics struct {
+	Labels               []string
```
done

akhurana001 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,229 @@
+	util.RegisterMetric(sparkAppExecutorSuccessCount)
+
+	return &sparkAppMetrics{
+		labels,
```
These are just the label names passed in while registering the metrics. The values are passed in to the exportMetrics function

liyinan926 replied:
I meant you should construct this `sparkAppMetrics` in the following form:
```
return &sparkAppMetrics{
  labels: labels,
  prefix: prefix,
  sparkAppSubmitCount: sparkAppSubmitCount,
  ...
```

This is much more robust than rely in the right order of the fields when initializing such struct.

liyinan926 commented on docs/quick-start-guide.md:
````diff
@@ -50,6 +50,43 @@ $ kubectl describe deployment sparkoperator -n sparkoperator
+
+```bash
+-enable-metrics=true
+-metrics-port=10254
````
It's worth mentioning that if users use a different port, they also need to update the annotation `prometheus.io/port` and `containerPort` in the yaml.

akhurana001 replied:
sure, done

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -501,20 +507,33 @@ func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdat
+
+		// Let's keep the old App status if this is not a valid transition.
+		if !isValidDriverStateTransition(original.Status.AppState.State, toUpdate.Status.AppState.State) {
+			toUpdate.Status.AppState = original.Status.AppState
```
We might need to log a `warning` here with the old and new state.

akhurana001 replied:
done, just FYI, this will be pretty common since same state transitions are also invalid. e.g. we will log going from RUNNING -> RUNNING.

liyinan926 replied:
Oh, right, I forgot that transition. Let's add another condition before the check for validity like the following:
```
if toUpdate.Status.AppState.State != original.Status.AppState.State && !isValidDriverStateTransition(original.Status.AppState.State, toUpdate.Status.AppState.State)
```

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/controller.go:
```diff
@@ -501,20 +507,33 @@ func (c *Controller) processSingleExecutorStateUpdate(update *executorStateUpdat
 	var lastUpdateErr error
 	for i := 0; i < maximumUpdateRetries; i++ {
-		updated, err := c.tryUpdateStatus(original, toUpdate, updateFunc)
+		toUpdate := original.DeepCopy()
```
Since the client always return a copy, the original approach that makes a copy once outside the loop and assign what the client returns to `toUpdate` is more desirable.

akhurana001 replied:
As I mentioned, this works fine if we aren't publishing metrics.

For metrics as well as to verify if the stateTransition is valid, we need to maintain the last returned by crdClient. 
We can set `toUpdate` as the last-state returned by the crdClient but then in the next attempt, we also do `	updateFunc(&toUpdate.Status)`  at https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/pull/227/files/f600336b353287b0cd983148071c2bf9640b1135#diff-b7eea120d7ac0e8587781b863b7db8e4R516 . At this point, we lose the last state returned by the `crdClient`. 

For us to  validate `isValidDriverStateTransition` and post metrics, we need to maintain that last state returned by the `crdClient` which is what `original` will be at this point ? 


I might be missing something but this is my understanding of the logic here. I be happy to jump on a quick call if need be.

liyinan926 replied:
Makes sense.

liyinan926 commented on pkg/controller/sparkapplication/controller_test.go:
```diff
@@ -287,10 +297,17 @@ func TestOnDelete(t *testing.T) {
 }
 
 func TestProcessSingleDriverStateUpdate(t *testing.T) {
+	type metrics struct {
```
Looks like the same type is used and defined in `TestProcessSingleAppStateUpdate` below, and a similar type is defined in `TestProcessSingleExecutorStateUpdate ` for executors. I think we can define a single type at the package level for this and use it for all three tests.

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,235 @@
+}
+
+func newSparkAppMetrics(prefix string, labels []string) *sparkAppMetrics {
+	for i, label := range labels {
```
To be safe, let's make and use a copy of `labels` here. The in-place changes to `labels` here are visible outside the caller of this function.

akhurana001 replied:
done

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,187 @@
+
+// Gauge with conditional decrement ensuring its value is never negative.
+func NewPositiveGauge(name string, description string, labels []string) *PositiveGauge {
+	for i, label := range labels {
```
Ditto. Let's make and use a copy of `labels`.

akhurana001 replied:
done

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,235 @@
+				if m, err := sm.sparkAppFailureExecutionTime.GetMetricWith(metricLabels); err != nil {
+					glog.Errorf("Error while exporting metrics: %v", err)
+				} else {
+					m.Observe(float64(d / time.Microsecond))
```
Can we do `float64(d) / time. Microsecond` instead here?

akhurana001 replied:
incompatible types: time.Microsecond is duration

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics.go:
```diff
@@ -0,0 +1,235 @@
+				if m, err := sm.sparkAppSuccessExecutionTime.GetMetricWith(metricLabels); err != nil {
+					glog.Errorf("Error while exporting metrics: %v", err)
+				} else {
+					m.Observe(float64(d / time.Microsecond))
```
Can we do `float64(d) / time. Microsecond` instead here?

akhurana001 replied:
same as above

liyinan926 commented on pkg/controller/sparkapplication/sparkapp_metrics_test.go:
```diff
@@ -0,0 +1,62 @@
+
+func TestSparkAppMetrics(t *testing.T) {
+	http.DefaultServeMux = new(http.ServeMux)
+	// Invalid label.
```
What does this comment mean?

akhurana001 replied:
So prometheus doesn't support "-" in metric or label names. Since we were specifying "-" in our workQueue names, I added a helper method to convert "-" to "_" and not error out (https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/pull/227/files#diff-48f67873a59be740236d811d032ce6f6R33) .. This test just validates that we are correctly handling that case as well.

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,187 @@
+	"strings"
+	"sync"
+
+	"fmt"
```
This import should be at the top.

akhurana001 replied:
sure, i have been running `goimports` on all files.

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,187 @@
+	MetricsLabels   []string
+}
+
+type PositiveGauge struct {
```
Can we add a doc for this like `PositiveGauge` is a variant of Prometheus Gauge that only holds non-negative values. BTW: this can be zero, right?

akhurana001 replied:
yes, this can be zero. Added comment

liyinan926 commented on pkg/util/metrics.go:
```diff
@@ -0,0 +1,187 @@
+	// Decrement only if positive
+	val := fetchGaugeValue(c.gaugeMetric, labelMap)
+	if val > 0 {
+		glog.V(2).Infof("Decrementing %s with labels %s metricVal to %v", c.name, labelMap, val-1)
```
Looks like indention is not consistent: 2 in some places, whereas 4 in other places. Please run `go fmt ./...` to format the code.

akhurana001 replied:
done

akhurana001 commented:
Thanks @liyinan926 . Please also update the image in https://github.com/GoogleCloudPlatform/spark-on-k8s-operator/blob/master/manifest/spark-operator-with-metrics.yaml#L48 once you build the operator with metrics.

liyinan926 commented:
Will do.
